deepchem/models/tensorflow_models/IRV.py

```diff
@@ -2,32 +2,142 @@
+       https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2750043/
+  """
+
+  def __init__(self, n_tasks, K, **kwargs):
+    """
+    Parameters
+    ----------
+    n_tasks: int
+      Number of tasks
+    K: int
+      Number of nearest neighbours used in classification
+    """
+    self.n_tasks = n_tasks
+    self.K = K
+    self.V, self.W, self.b, self.b2 = None, None, None, None
+    super(IRVLayer, self).__init__(**kwargs)
+
+  def build(self):
+    self.V = tf.Variable(tf.constant([0.01, 1.]), name="vote", dtype=tf.float32)
+    self.W = tf.Variable(tf.constant([1., 1.]), name="w", dtype=tf.float32)
+    self.b = tf.Variable(tf.constant([0.01]), name="b", dtype=tf.float32)
+    self.b2 = tf.Variable(tf.constant([0.01]), name="b2", dtype=tf.float32)
+    self.trainable_weights = [self.V, self.W, self.b, self.b2]
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    if in_layers is None:
+      in_layers = self.in_layers
+    in_layers = convert_to_layers(in_layers)
+
+    self.build()
+    inputs = in_layers[0].out_tensor
+    K = self.K
+    outputs = []
+    for count in range(self.n_tasks):
+      # Similarity values
+      similarity = inputs[:, 2 * K * count:(2 * K * count + K)]
+      # Labels for all top K similar samples
+      ys = tf.to_int32(inputs[:, (2 * K * count + K):2 * K * (count + 1)])
+
+      R = self.b + self.W[0] * similarity + self.W[1] * tf.constant(
+          np.arange(K) + 1, dtype=tf.float32)
+      R = tf.sigmoid(R)
+      z = tf.reduce_sum(R * tf.gather(self.V, ys), axis=1) + self.b2
+      outputs.append(tf.reshape(z, shape=[-1, 1]))
+    out_tensor = tf.concat(outputs, axis=1)
+
+    if set_tensors:
+      self.variables = self.trainable_weights
+      self.out_tensor = out_tensor
+    return out_tensor
+
+  def none_tensors(self):
+    V, W, b, b2 = self.V, self.W, self.b, self.b2
+    self.V, self.W, self.b, self.b2 = None, None, None, None
+
+    out_tensor, trainable_weights, variables = self.out_tensor, self.trainable_weights, self.variables
+    self.out_tensor, self.trainable_weights, self.variables = None, [], []
+    return V, W, b, b2, out_tensor, trainable_weights, variables
+
+  def set_tensors(self, tensor):
+    self.V, self.W, self.b, self.b2, self.out_tensor, self.trainable_weights, self.variables = tensor
+
+
+class IRVRegularize(Layer):
+  """ This Layer extracts the trainable weights in IRVLayer
+  and return the their L2-norm
+  """
+
+  def __init__(self, IRVLayer, penalty=0.0, **kwargs):
+    self.IRVLayer = IRVLayer
+    self.penalty = penalty
+    super(IRVRegularize, self).__init__(**kwargs)
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    assert self.IRVLayer.out_tensor is not None, "IRVLayer must be built first"
+    out_tensor = tf.nn.l2_loss(self.IRVLayer.W) + \
+        tf.nn.l2_loss(self.IRVLayer.V) + tf.nn.l2_loss(self.IRVLayer.b) + \
+        tf.nn.l2_loss(self.IRVLayer.b2)
+    out_tensor = out_tensor * self.penalty
```

Comment:
IRV has l2 regularization, is there a better way to handle these extra loss values?

Reply:
I think this looks fine.